[YUNIKORN-971] Implement YuniKorn as a Kubernetes scheduler plugin.

craigcondit opened the pull request:
### What is this PR for?
This PR implements a new build mode for YuniKorn which allows it to run as a set of plugins on top of the default Kubernetes scheduler. This should allow for closer behavior to the default scheduler as well as the ability to bypass YuniKorn-specific functionality for pods which are not annotated with an `applicationId` label.

This PR makes no functional changes to the default YuniKorn build, but adds new Makefile targets to support build / running YuniKorn as a scheduler plugin:

- `build_plugin` Build the plugin code.
- `run_plugin` Run the plugin code locally.
- `plugin_image` Create a Docker image from the plugin code. this should mostly be a drop-in replacement for the default YuniKorn image.

This feature should be considered `experimental` at this stage. Hopefully we can provide it as a tech-preview for 1.0.

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [x] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-971

### How should this be tested?
Some unit tests added for configuration, remainder is covered via e2e tests (all passing).

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

craigcondit commented:
Rebased on latest master (post 0.12.1).

craigcondit commented:
@wilfred-s, @yangwwei, can you review?

pbacsko commented on pkg/cache/external/scheduler_cache.go:
```diff
@@ -36,23 +36,23 @@ import (
+	nodesMap              map[string]*framework.NodeInfo // node name to NodeInfo map
+	podsMap               map[string]*v1.Pod
+	assumedPods           map[string]bool   // map of assumed pods, value indicates if pod volumes are all bound
+	pendingAllocations    map[string]string // map of pod to node ID, presence indicates a pending allocation for scheduler
+	inProgressAllocations map[string]string // map of pod to node ID, presence indicates an in-process allocation for scheduler
```
We need some extra comment which explains what is the difference between a "pending" and an "in progress" allocation. Just by looking at the code, it's not straightforward which comes first and what is the lifecycle of the pods.

craigcondit replied:
A pending allocation is one which YuniKorn has decided on, but which has not yet been communicated to the default scheduler via the scheduling framework (the PreFilter()/Filter() calls in the plugin). An in-progress allocation is one which HAS been communicated to the default scheduler via PreFilter() / Filter() but which we have not yet got confirmation of the Pod being bound (PostBind() in the plugin has not yet been called). This is used to track state to ensure that we don't reprocess the same Pod or leak in the case where the default scheduler is unable to bind the pod for some reason.

craigcondit replied:
Added additional documentation in scheduler_cache.go.

pbacsko replied:
Thanks, this definitely helps to understand the code.

pbacsko commented on pkg/cache/task.go:
```diff
@@ -440,6 +460,23 @@ func (task *Task) beforeTaskAllocated(event *fsm.Event) {
+		// queue for unschedulableQTimeInterval long (default 1 minute). hence, we are updating
+		// the pod status explicitly, when there is a status change, the default scheduler will
+		// move the pod back to the active queue immediately.
+		podCopy := task.pod.DeepCopy()
```
Can we reason about the performance effects of this copy? If we do this all the time (and that's what seems to be happening), we end up creating extra objects which affects GC time. Maybe not a big deal, just something to think about.

craigcondit replied:
This cost is negligible, because the copy happens very infrequently (typically once per pod), when YuniKorn has decided to allocate the pod. We update the pod status once at this point to force the default scheduler logic to re-consider the pod for scheduling. It will very shortly call PreFilter() / Filter() and push the Pod through the scheduling cycle.

pbacsko commented on pkg/cache/external/scheduler_cache.go:
```diff
@@ -132,6 +132,45 @@ func (cache *SchedulerCache) removeNode(node *v1.Node) error {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+	expectedNodeID, ok := cache.pendingAllocations[podKey]
+	if ok && expectedNodeID == nodeID {
```
Ok, from what I understand `expectedNodeID` comes from YK itself, and `nodeID` from the default scheduler. Questions:
* How often do these two mismatch?
* What happens if we return `false` here?
* Can YK and the def. scheduler end up repeatedly allocating to different nodes? 

Another thing is that do we even need to consider YK's allocation? Because in `task.postTaskAllocated()`, the original code path calls `KubeClient.Bind()` but in plugin mode, we delegate this operation as I can see.

craigcondit replied:
This code is used to verify that the requested node (by the default scheduler) is the one that YuniKorn decided to allocate the pod to. Returning `false` here will result in the Filter() call in the plugin returning `Unschedulable` to the default scheduler for this pod / node combination, and a subsequent Filter() call with a matching node will return `Success` instead. Effectively, we use this code to veto node allocations other than the node that YK has decided on.

The bind is deferred to the default scheduler logic once Filter() succeeds, which is why we don't handle the bind logic in postTaskAllocated().

pbacsko commented on pkg/schedulerplugin/scheduler_plugin.go:
```diff
@@ -0,0 +1,192 @@
+
+			nodeID, ok := sp.context.GetPendingPodAllocation(string(pod.UID))
+			if !ok {
+				nodeID = "<none>"
```
Is this value used for anything?

craigcondit replied:
in the logging message below to make things clearer.

craigcondit replied:
Per offline discussion, removed this as the log message will never print it anyway.

pbacsko commented on pkg/cache/external/scheduler_cache.go:
```diff
@@ -132,6 +132,45 @@ func (cache *SchedulerCache) removeNode(node *v1.Node) error {
+func (cache *SchedulerCache) AddPendingPodAllocation(podKey string, nodeID string) {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+	delete(cache.inProgressAllocations, podKey)
```
Is this delete necessary? Under what circumstances can we have an entry?

As I can see it, `inProgressAllocations` is populated from `StartPodAllocation()` but that already requires that pending has an entry which we just add in line 139. Am I missing something here?

Is it a bug if we already have something in `inProgressAllocations` at this point for the same `podKey`?

craigcondit replied:
If a pod fails to bind in the default scheduler for some reason, the only notice we may get (as there is not a callback in the scheduler framework for failed binds) is another call to PreFilter() / Filter() for this pod. So it is possible in that case for us to have an in-progress allocation. So, we remove it and reset the state back to pending to be safe.

craigcondit commented:
Changes in latest version:

- Rebased PR on master
- Fixed event recorder changes after YUNIKORN-937
- Implemented `--plugin` flag for `scripts/run-e2e-tests.sh`. E2E tests can be run using the plugin implementation with something like:

`scripts/run-e2e-tests.sh -a test -n "yk8s" -v "kindest/node:v1.21.2" --plugin`

craigcondit commented:
Rebased on master.

craigcondit commented:
Added additional documentation to the scheduler_plugin.go and scheduler_cache.go files. Hopefully this is now clear.

craigcondit commented:
@pbacsko can you review again?

pbacsko commented on pkg/schedulerplugin/scheduler_plugin.go:
```diff
@@ -0,0 +1,209 @@
+	return nil
+}
+
+// Filter is used to release specific pod/node combinations to scheduler
```
The thing that's probably worth mentioning here is what happens when `expectedNode != node`, pointing out the minimal performance impact and the immediate invocation of `Filter()` again. You can probably also add that currently this is the only way to force the default scheduler to match YK's decision.

craigcondit replied:
Added comments.

pbacsko commented on pkg/schedulerplugin/scheduler_plugin.go:
```diff
@@ -0,0 +1,209 @@
+	SchedulerPluginName = "YuniKornPlugin"
+)
+
+// YuniKornSchedulerPlugin provides an implementation of several lifecycle methods of the Kubernetes scheduling framework:
+//   https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/
+//
+// PreFilter: Used to notify the default scheduler that a particular pod has been marked ready for scheduling by YuniKorn
+//
+// Filter: Used to notify the default scheduler that a particular pod/node combination is ready to be scheduled
+//
+// PostBind: Used to notify YuniKorn that a pod has been scheduled successfully
+//
+// Pod Allocations:
+//
+// The YuniKorn scheduler is always running in the background, making decisions about which pods to allocate to which
+// nodes. When a decision is made, that pod is marked as having a "pending" pod allocation, which means YuniKorn has
+// allocated the pod, but the default scheduler (via the plugin interface) has not yet been notified.
+//
+// Once PreFilter() has been called for a particular pod, that allocation is marked as "in progress" meaning it has been
+// communicated to the default scheduler, but has not yet been fulfilled.
+//
+// Finally, in PostBind(), the allocation is removed as we now know that the pod has been allocated successfully.
+// If a pending or in-progress allocation is detected for a pod in PreFilter(), we remove the allocation and force the
+// pod to be rescheduled, as this means the prior allocation could not be completed successfully by the default
+// scheduler for some reason.
+type YuniKornSchedulerPlugin struct {
```
Solid comment, it's great to have this here.

pbacsko reviewed (changes requested):
Pending LGTM +1 - I just asked to add one more minor thing.

craigcondit commented:
@pbacsko can you re-review? I've addressed the comments you made on the last revision.

pbacsko reviewed (approved):
+1
